optional_plugins/golang/avocado_golang/__init__.py

```diff
@@ -0,0 +1,183 @@
+
+
+class NotGolangTest(object):
+
+    """
+    Not a golang test (for reporting purposes)
+    """
+
+
+class GolangLoader(loader.TestLoader):
+    """
+    Golang loader class
+    """
+    name = "golang"
+
+    def __init__(self, args, extra_params):
+        super(GolangLoader, self).__init__(args, extra_params)
+
+    def discover(self, url, which_tests=loader.DEFAULT):
+        if url is None:
+            return []
+
+        avocado_suite = []
+        package_paths = []
+        test_files = []
+        subtest = None
+        tests_filter = None
+        go_root = os.environ.get('GOROOT')
+        go_path = os.environ.get('GOPATH')
+
+        if go_root is not None:
+            package_paths.append(os.path.join(os.path.expanduser(go_root),
+                                              'src'))
+        if go_path is not None:
+            package_paths.append(os.path.join(os.path.expanduser(go_path),
+                                              'src'))
+
+        if ':' in url:
+            url, _tests_filter = url.split(':', 1)
+            if '/' in _tests_filter:
+                _tests_filter, subtest = _tests_filter.split('/', 1)
+            tests_filter = re.compile(_tests_filter)
+
+        for package_path in package_paths:
+            url_path = os.path.join(package_path, url)
+            files = self._find_files(url_path)
+            if files:
+                test_files.append((package_path, files))
+                break
+
+        if not test_files:
+            if which_tests == loader.ALL:
+                msg = ('No package matching this reference.')
+                return [(NotGolangTest,
+                         {"name": "%s: %s" % (url, msg)})]
+            return []
+
+        for package_path, test_files_list in test_files:
+            for test_file in test_files_list:
+                test_suite = self._find_tests(test_file)
+                for item in test_suite:
+                    common_prefix = os.path.commonprefix([package_path,
+                                                          test_file])
+                    match_package = os.path.relpath(test_file, common_prefix)
+                    test_name = "%s:%s" % (os.path.dirname(match_package),
+                                           item)
+                    if tests_filter and not tests_filter.search(test_name):
+                        continue
+                    avocado_suite.append((GolangTest,
+                                          {'name': test_name,
+                                           'subtest': subtest}))
+        return avocado_suite
+
+    @staticmethod
+    def _find_tests(test_path):
+        test_suite = []
+        with open(test_path, 'r') as test_file_fd:
+            for line in test_file_fd.readlines():
+                if line.startswith('func Test'):
+                    test_suite.append(line.split()[1].split('(')[0])
+
+        return test_suite
+
+    @staticmethod
+    def _find_files(path):
```

Comment:
This should allow passing individual files, something like this fixes it:

```diff
diff --git a/optional_plugins/golang/avocado_golang/__init__.py b/optional_plugins/golang/avocado_golang/__init__.py
index 6a806d0..3c0f54b 100644
--- a/optional_plugins/golang/avocado_golang/__init__.py
+++ b/optional_plugins/golang/avocado_golang/__init__.py
@@ -149,12 +149,14 @@ class GolangLoader(loader.TestLoader):
 
     @staticmethod
     def _find_files(path):
-        matches = []
-        for root, dirnames, filenames in os.walk(path):
-            for filename in fnmatch.filter(filenames, '*_test.go'):
-                matches.append(os.path.join(root, filename))
-
-        return matches
+        if os.path.isdir(path):
+            matches = []
+            for root, dirnames, filenames in os.walk(path):
+                for filename in fnmatch.filter(filenames, '*_test.go'):
+                    matches.append(os.path.join(root, filename))
+            return matches
+        elif os.path.exists(path):
+            return [path]
 
     @staticmethod
     def get_type_label_mapping():
```

Reply:
Ok, I will include the individual files or directories.
Some issues can happen when using absolute paths instead of packages, but on, let's allow that and let the user deal with the problems, the same way they have to deal if using `go test` directly.